superset/views/tags.py

```diff
@@ -0,0 +1,197 @@
+
+
+def process_template(content):
+    env = SandboxedEnvironment()
+    template = env.from_string(content)
+    context = {
+        'current_user_id': current_user_id,
+        'current_username': current_username,
+    }
+    return template.render(context)
+
+
+def get_name(obj):
+    if obj.Dashboard:
+        return obj.Dashboard.dashboard_title
+    elif obj.Slice:
+        return obj.Slice.slice_name
+    elif obj.SavedQuery:
+        return obj.SavedQuery.label
+
+
+def get_creator(obj):
+    if obj.Dashboard:
+        return obj.Dashboard.creator()
+    elif obj.Slice:
+        return obj.Slice.creator()
+    elif obj.SavedQuery:
+        return obj.SavedQuery.creator()
+
+
+def get_attribute(obj, attr):
+    if obj.Dashboard:
+        return getattr(obj.Dashboard, attr)
+    elif obj.Slice:
+        return getattr(obj.Slice, attr)
+    elif obj.SavedQuery:
+        return getattr(obj.SavedQuery, attr)
+
+
+class TagView(BaseSupersetView):
+
+    @expose('/tags/suggestions/', methods=['GET'])
+    def suggestions(self):
+        query = db.session.query(
+            TaggedObject,
+        ).group_by(TaggedObject.tag_id).order_by(func.count().desc()).all()
+        tags = json.dumps([
+            {'id': obj.tag.id, 'name': obj.tag.name} for obj in query])
+
+        return Response(tags, status=200, content_type='application/json')
+
+    @expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['GET'])
+    def get(self, object_type, object_id):
+        """List all tags a given object has."""
+        query = db.session.query(TaggedObject).filter(and_(
+            TaggedObject.object_type == object_type,
+            TaggedObject.object_id == object_id))
+        tags = json.dumps([
+            {'id': obj.tag.id, 'name': obj.tag.name} for obj in query])
+
+        return Response(tags, status=200, content_type='application/json')
+
+    @expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['POST'])
+    def post(self, object_type, object_id):
+        """Add new tags to an object."""
+        tagged_objects = []
+        for name in request.get_json(force=True):
+            if ':' in name:
+                type_name = name.split(':', 1)[0]
+                type_ = TagTypes[type_name]
+            else:
+                type_ = TagTypes.custom
+
+            tag = db.session.query(Tag).filter_by(name=name, type=type_).first()
+            if not tag:
+                tag = Tag(name=name, type=type_)
+
+            tagged_objects.append(
+                TaggedObject(
+                    object_id=object_id,
+                    object_type=object_type,
+                    tag=tag,
+                ),
+            )
+
+        db.session.add_all(tagged_objects)
+        db.session.commit()
+
+        return Response(status=201)  # 201 CREATED
+
+    @expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['DELETE'])
+    def delete(self, object_type, object_id):
```

Comment:
Note that some of these endpoints we'd get for free by creating a FAB ModelView object, it may not be exactly what you need though, unclear. I'm ok with the more explicit/controlled endpoints.

Reply:
👍